fix(*): Helm v3 handling of APIVersion v1 charts dependencies

TBBle opened the pull request:
This fixes unexpected changes in APIVersion1 (Helm v2) Charts when they are operated on by Helm v3 using `package` or `dependency update`.

The fixes:
- `requirements.yaml` was being unconditionally merged into `Chart.yaml`, which produced packages which failed `helm lint`.
- `requirements.lock` was being converted into `Chart.lock` by `helm dep up` on an APIVersion1 chart.
- The code that loads `requirements.lock` and `requirements.yaml` in APIVersion1 charts for backwards compatibility, was incorrectly excluding those files from `helm package` and the `Files` collection in templates. As well as `Files` behaviour changing, this also broke sub-chart aliases for Helm v2 as those are stored in `requirements.yaml`.

Helm v3 incorrectly reads aliases from `Chart.yaml`'s `dependencies` map in APIVersion1 charts, but I didn't change that, as `helm lint` warns you that this is an unexpected combination already, and it's *as well as* `requirements.yaml` so doesn't break anything.

The changes in the third commit (ae74c91) are possibly not go-idiomatic, and I specifically welcome feedback or suggested improvements.

Fixes: #6974

Commit:
Include requirements.* as Files in APIVersionV1

Fixes #6974.

This ensures that when reading a Chart marked with APIVersion v1, we
maintain the behaviour of Helm v2 and include the requirements.yaml and
requirements.lock in the Files collection, and hence produce charts that
work correctly with Helm v2.

Signed-off-by: Paul "Hampy" Hampson <p_hampson@wargaming.net>

Commit:
Write out requirements.lock for APIVersion1 Charts

This keeps the on-disk format consistent after `helm dependency update`
of an APIVersion1 Chart.

Signed-off-by: Paul "Hampy" Hampson <p_hampson@wargaming.net>

Commit:
Exclude 'dependencies' from APVersion1 Chart.yaml

This fixes `helm lint` against an APIVersion1 chart packaged with Helm
v3.

Signed-off-by: Paul "Hampy" Hampson <p_hampson@wargaming.net>

TBBle commented:
I intend to fix the code-coverage failure, but not immediately. Hopefully later this week, pending other workload.

Commit:
Generate APIVersion v2 charts for dependency tests

As the generated chart contains no requirements.yaml in its files list,
but has dependencies in its metadata, it is not a valid APIVersion v1
chart.

Signed-off-by: Paul "Hampy" Hampson <p_hampson@wargaming.net>

TBBle commented:
I haven't added tests to cover the APIVersion v1 behaviours; the changes were visible in the test runs as they broke the dependency tests which were testing APIVersion v2 behaviours with a generated APIVersion v1 chart. Instead I changed those tests to test the behaviour of APIVersion v2 charts, and they pass again.

TBBle commented:
Is there any kind of process to go through to have a PR reviewed? This's been sitting idle for a couple of weeks now, and I'd love to know if there's more I need to do.

hickeyma commented on cmd/helm/dependency_update_test.go:
```diff
@@ -182,7 +182,7 @@ func TestDependencyUpdateCmd_DontDeleteOldChartsOnError(t *testing.T) {
 	return &chart.Chart{
 		Metadata: &chart.Metadata{
-			APIVersion: chart.APIVersionV1,
+			APIVersion: chart.APIVersionV2,
```
Can you make this change also in the following places:
- https://github.com/helm/helm/blob/master/pkg/downloader/manager_test.go#L214
- https://github.com/helm/helm/blob/master/pkg/downloader/manager_test.go#L288

hickeyma commented on pkg/chartutil/chartfile.go:
```diff
@@ -42,7 +42,16 @@ func LoadChartfile(filename string) (*chart.Metadata, error) {
 //
 // 'filename' should be the complete path and filename ('foo/Chart.yaml')
 func SaveChartfile(filename string, cf *chart.Metadata) error {
+	// Pull out the dependencies of a v1 Chart, since there's
```
You can exclude fields in the serializer but it is at definition time and would therefore exclude the field from all serializing (e.g. `Dependencies []*Dependency json:"dependencies,omitempty"` --> `Dependencies []*Dependency`). This seems best option as it is just for the v1 case and not in all cases.

hickeyma replied:
Maybe update the comment

TBBle replied:
I'm not sure what you want changed in the comment. Would it be clearer to say "there's no way to tell the serialiser to conditionally skip a field"?

TBBle replied:
> > requirements.yaml was being unconditionally merged into Chart.yaml, which produced packages which failed helm lint.
> 
> I am unable to reproduce this. Can you give more details about it?

With Helm v2.16.2:
```powershell
helm create testchart
```

Add a requirements.yaml:
```yaml
dependencies:
  - name: magic-ip-address
    version: ">0"
    repository: "@stable"
```

With Helm v2.16.2:
```powershell
helm dependency update testchart
```

With Helm v3.0.0 or v3.0.1
```powershell
helm package testchart
helm lint testchart-0.1.0.tgz
```

This will reproduce the failure:
```
==> Linting .\testchart-0.1.0.tgz
[INFO] Chart.yaml: icon is recommended
[ERROR] Chart.yaml: dependencies are not valid in the Chart file with apiVersion 'v1'. They are valid in apiVersion 'v2'   
Error: 1 chart(s) linted, 1 chart(s) failed
```

hickeyma replied:
Maybe something like this: "no way to tell the serialiser to skip a field" --> "no way to tell the serialiser to skip a field for just this use case"

hickeyma commented on pkg/chartutil/save.go:
```diff
@@ -141,8 +141,17 @@ func Save(c *chart.Chart, outDir string) (string, error) {
 func writeTarContents(out *tar.Writer, c *chart.Chart, prefix string) error {
 	base := filepath.Join(prefix, c.Name())
 
+	// Pull out the dependencies of a v1 Chart, since there's
```
Same as comment in `pkg/chartutil/chartfile.go `

hickeyma reviewed (changes requested):
@TBBle Thank you for pushing this PR. It is a good fix. I have added some comments for consideration.  
One other change is that the `requirements.lock` file should not be saved when packaging. The check could potentially be added before: https://github.com/helm/helm/blob/master/pkg/chartutil/save.go#L182

hickeyma commented:
@TBBle Thanks for the details. Some additional feedback below.

> requirements.yaml was being unconditionally merged into Chart.yaml, which produced packages which failed helm lint.

I am unable to reproduce this. Can you give more details about it?

> The changes in the third commit (ae74c91) are possibly not go-idiomatic, and I specifically welcome feedback or suggested improvements.

Added feedback in https://github.com/helm/helm/pull/7009#discussion_r355539730

TBBle commented:
Why should `requirements.lock` be excluded from packaging? It was included in packages generated by Helm v2.

TBBle commented:
> > requirements.yaml was being unconditionally merged into Chart.yaml, which produced packages which failed helm lint.
> 
> I am unable to reproduce this. Can you give more details about it?

With Helm v2.16.2:
```powershell
helm create testchart
```

Add a requirements.yaml:
```yaml
dependencies:
  - name: magic-ip-address
    version: ">0"
    repository: "@stable"
```

With Helm v2.16.2:
```powershell
helm dependency update testchart
```

With Helm v3.0.0 or v3.0.1
```powershell
helm package testchart
helm lint testchart-0.1.0.tgz
```

This will reproduce the failure:
```
==> Linting .\testchart-0.1.0.tgz
[INFO] Chart.yaml: icon is recommended
[ERROR] Chart.yaml: dependencies are not valid in the Chart file with apiVersion 'v1'. They are valid in apiVersion 'v2'   
Error: 1 chart(s) linted, 1 chart(s) failed
```

hickeyma commented:
> Why should requirements.lock be excluded from packaging of APIVersion1 charts? It was included in packages generated by Helm v2, reconfirmed with Helm v2.16.2 with the same setup as the following repro case

I agree that it is included in Helm 2 but the `lock` file is not included in Helm 3. Is it still needed in the package as you will have the dependencies included? I would probably like to mirror Helm 3 in this regard.

hickeyma commented:
> requirements.yaml was being unconditionally merged into Chart.yaml, which produced packages which failed helm lint.

Thanks for clarification of this in https://github.com/helm/helm/pull/7009#issuecomment-563989554.  I missed the subtly of doing update first with Helm 2.

TBBle commented:
> > Why should requirements.lock be excluded from packaging of APIVersion1 charts? It was included in packages generated by Helm v2, reconfirmed with Helm v2.16.2 with the same setup as the following repro case

> I agree that it is included in Helm 2 but the `lock` file is not included in Helm 3. Is it still needed in the package as you will have the dependencies included? I would probably like to mirror Helm 3 in this regard.

Personally, I'd keep it around for API stability reasons, in case someone's scripts are using `.Files` to extract the real subchart versions from `requirements.lock` or similar. It seems a reasonably-arbitrary API break for `v1` to suddenly disappear the file.

Commit:
Generate APIVersion v2 charts for manager tests

Specifically for the charts that have dependencies, the generated chart
contains no requirements.yaml in its files but has dependencies in its
metadata. Hence it is not a valid APIVersion v1 chart.

Signed-off-by: Paul "Hampy" Hampson <p_hampson@wargaming.net>

hickeyma reviewed (approved):
LGTM, thanks for the fix @TBBle

technosophos commented:
picked into release-3.0

Nitive commented:
@TBBle generating `requirements.lock` instead of `Chart.lock` was a breaking change for helmfile, you can see details in https://github.com/roboll/helmfile/issues/1052. Not sure if it should be fixed on a helm side

TBBle commented:
As noted in roboll/helmfile#1052, the change back to requirements.yaml for Helm APIVersion 1 charts was reverting an incorrect API breakage, and has now been fixed in helmfile, which was incorrectly generating APIVersion 1 (by default) charts with the APIVersion 2 `Chart.lock`; so were the unit tests in Helm v3 itself until this fix, so it's not surprising, just a shame this wasn't caught earlier in the Helm v3 release cycle.
